### PR TITLE
Site Assembler - Sections Pattern Inserter, Categories, and Large Preview UX

### PR DESCRIPTION
**Pull request by miksansegundo**
Related to #75327

## Proposed Changes

- The "Homepage" screen won't show if there aren't patterns added. The user goes directly to the "Add patterns" screen
- The "Add patterns" screen has the first category open by default
- Allow toggling the category to hide the patterns panel
- The large preview is fully visible and the patterns panel doesn't float over it

## Testing Instructions

- Click on the Calypso Live link in the first comment
- Create a new site with any plan
- Continue until the Design Picker
- Scroll down and click `Start designing`
- Click `Homepage` and verify
  - The  `Homepage` screen only shows when there are patterns selected
  - The `Back` and `Save` buttons go to the previous screen
- Verify that the large preview is fully visible
- Verify you can switch devices and scroll the large preview
- Verify you can click on a category to toggle the patterns panel
- Test patterns from the "Link in bio" category to test that they use the 100% viewport height 

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by lucasmendes-design** [comment]
Great work. 

**Some notes:** 
- I'm worried about how it will look on small screens.  
- I can't deselect the category anymore. It's always selected and expanded. The video below is the comparison between this and in-production. Not sure if it was intentional but I think we could collapse the categories. It helps for small screens. WDYT? 

https://github.com/Automattic/wp-calypso/assets/98343583/ca717419-3d78-4196-9cba-94c02477252c

**Comment by miksansegundo** [comment]
> I'm worried about how it will look on small screens.

Me too. I think the preview on the desktop viewport (when we scale it down), could look too small.

> I can't deselect the category anymore. It's always selected and expanded. 

This is intentional. The reasoning is that the `Add patterns` screen goal is adding patterns and previewing them (fully) in the large preview. When you are done adding patterns, you click save/back and the large preview will get expanded again for you to continue working on your homepage on a bigger canvas.

I get your point, a user on a small screen could want to make the large preview larger. As a user, what would motivate you to click on the same category to toggle? I think that I would click if I see the patterns panel floating over. 

One idea, is to make the assembler responsive to smaller screens with screen breakpoints to adjust paddings.

**Comment by autumnfjeld** [comment]
Just an FYI: This is what happens in the editor when the browser window is narrow:

https://github.com/Automattic/wp-calypso/assets/5665959/4f2c4413-521e-479d-a7d0-5c49e93c41df




What was your browser viewport width in your demo @miksansegundo ? Full laptop monitor width?

**Comment by autumnfjeld** [comment]
@miksansegundo @taipeicoder 
My two cents are, that if we made any design/dev effort to change the viewing/responsive/etc for "small screens" that this is its own project and  we don't move forward on it until all the work in pekYwv-1Df-p2   is done and we have a mini CfT or internal CfT on this work. Then we see if anyone comments on the small screen. Would want to know a dev effort estimate before deciding to do this work.

Because the user can close the pattern selection panel, meaning after they click save, they will see their homepage in full view.   So maybe there should be some informative copy about that? Not sure. But my point is the user has the ability to see their homepage in a "proper" width once they are out of the pattern selection state. 

@lucasmendes-design  expressed similar ideas:
> I can't deselect the category anymore. It's always selected and expanded. The video below is the comparison between this and in-production. Not sure if it was intentional but I think we could collapse the categories. It helps for small screens. WDYT?

**Comment by autumnfjeld** [comment]
It might be helpful for the user to have a way to close the pattern panel.   Close the patten panel, so only the category list shows. Allowing more flexibility in viewing options while user is selecting patterns.

image

**Comment by miksansegundo** [comment]
> Allowing more flexibility in viewing options while user is selecting patterns.

@autumnfjeld what do you mean by viewing options? If the pattern panel is closed, users cannot add patterns.

I think when adding patterns they should be visible. I think users will only find it helpful to close the panel to make the large preview larger when it's too small but I don't see how it gives more flexibility to add patterns. 

@lucasmendes-design happy to explore where to put a button to close the panel if you also see the need for that feature.

**Comment by miksansegundo** [comment]
> What was your browser viewport width in your demo @miksansegundo ? Full laptop monitor width?

It was just a portion of my 39.5-inch panoramic screen 😅

Here is a full screenshot of a 16-inch laptop:
Screenshot 2566-05-31 at 12 04 26

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx]: [nit] Maybe `hasSections` is enough. WDYT?

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss:20]: I guess we have to increase the margin to balance the whitespace on both sides of the large preview and keep the box-shadow. WDYT?

**Comment by arthur791004** [comment]
I'd also prefer to toggle the pattern panel list 😆 We can allow people to unselect the category or have a `X` button on that panel to close it.

For the small screen, I think the pattern panel list should be on the top of the sidebar instead of growing on the right side. Here is the example in the site editor, and you can see the pattern list is on the top of the category list and there is a back button to allow users to go back to category list on the small screen

image

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.tsx
Comment (arthur791004) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.tsx:130]: I'd prefer to keep this feature to toggle the panel list. Let's wait for the final decision 🙂

Comment (miksansegundo) [reply client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.tsx:130]: Okay. Let's keep it.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss
Comment (miksansegundo) [reply client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss:20]: I agree about the margins. I think the shadow doesn't work well.
I did the changes following the design in Figma. Let's wait for Lucas' design review before updating that.

**Comment by miksansegundo** [comment]
A new version is ready for testing.

https://github.com/Automattic/wp-calypso/assets/1881481/a82ececf-7878-4120-9fe2-691299206a8d


Thank y'all for the feedback!

**Comment by lucasmendes-design** [comment]
Hi @miksansegundo, I tested and it looks good to me 😃 Thanks!

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss
Comment (arthur791004) [reply client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.scss:20]: Sure 🙂

**Comment by arthur791004** [comment]
> Scale to 1280px the desktop viewport

I think this behavior is a bit weird. For example, the viewport width is only fixed for the desktop device. If you switch to a tablet or mobile, the content will become responsive.

If we want to have a “fixed” viewport width for the device content, I assume we have to make the `DeviceSwithcer` component support this behavior so the viewport width will be fixed for all devices.

https://github.com/Automattic/wp-calypso/assets/13596067/52f2dbc3-199d-4ba4-b2c5-b7fc49c30078

**Comment by miksansegundo** [comment]
> If we want to have a “fixed” viewport width for the device content, I assume we have to make the DeviceSwithcer component support this behavior so the viewport width will be fixed for all devices.

👍 I agree with moving the feature to `DeviceSwithcer`, not sure if we necessarily have to fix the viewport width for all devices but we can test it.

**Comment by autumnfjeld** [comment]
> @autumnfjeld what do you mean by viewing options? If the pattern panel is closed, users cannot add patterns.

I meant all the different "states" the "views" the user has within the  PA. Example
* When selecting patterns, pattern panel open, the view is narrow
* When done selecting patterns, pattern panel is closed, the view is wide

Agree, when the pattern panel is closed, user can't add patterns. My point was more that the user can switch between these views, that the user does have options for a wider view, even if it is "effort" to switch between these views.

**Comment by autumnfjeld** [comment]
> 👍 I agree with moving the feature to DeviceSwithcer, not sure if we necessarily have to fix the viewport width for all devices but we can test it.

@miksansegundo @lucasmendes-design   Feels like this should be separate work, if it is needed at all.  Peopel will use the width of their screens in most cases, so that shoud be the design target.   Or maybe I'm missing something of why this would be worth the dev work?

**Comment by autumnfjeld** [comment]
@miksansegundo  Just tested the live link. Flow is a great improvement!!!   Nice job!  Might share your demo video in a comment of the P2, because this is such a news worthy improvement. 😄

**Comment by miksansegundo** [comment]
I think the desktop viewport width should be `1024 px minimum` and `ideally 1280 px`.

A Macboox 16" screen has `1728 x 1000 pixels` and the large preview gets only `963 px` **without scaling it,** which feels "okay" but it's still too small for a desktop.

Screenshot 2566-06-06 at 12 23 25


A Macboox 13" screen has `1280 x 800 pixels` and the large preview gets only `515 px` **without scaling it,** which feels like a tablet viewport. 

Screenshot 2566-06-06 at 12 21 36

We can address the scaling on another PR but I think they should be shipped together because `963 px` or less is not enough for the desktop preview.

cc: @autumnfjeld @lucasmendes-design

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx
Comment (okmttdhr) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx]: badge

I think we don't need this condition as it's already in `useEffect`!

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx
Comment (okmttdhr) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx]: badge

We might be able to create a variable for `205` to be consistent.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx
Comment (okmttdhr) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx]: badge

If we just want to execute a function in a certain condition, wouldn't it be ok to remove useEffect? (I'm sorry if it doesn't make sense here.)
https://react.dev/learn/you-might-not-need-an-effect

```suggestion
		if ( NAVIGATOR_PATHS.SECTION_PATTERNS === navigator.location.path ) {
			setTimeout(
				() => {
					updateViewportHeight();
					setLargePreviewStyleForViewport();
				},
				// Wait for .pattern-large-preview transition
				205
			);
		}
```

**Comment by arthur791004** [comment]
@miksansegundo I assume we can leave the viewport issue as it is (without scaling implemented in this PR) as the user encounters the same issue now when they are on the small screen. Let's make the `DeviceSwitcher` component support the fixed viewport width in the follow-up PR and move this forward. WDYT?

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx
Comment (miksansegundo) [reply client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx]: I noticed that the value `navigator.location` was not updated when is out of the effect. I'll check it again.

**Comment by miksansegundo** [comment]
@arthur791004 Okay, let's do that. I'll use this PR as the base for my next. Note that there are still some changes in this PR to fix issues. See my new comments on the code.

@okmttdhr Thanks for the review. I'll address the feedback in the next PR because I removed from this PR those changes.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx
Comment (miksansegundo) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx:175]: `true` makes it a passive event for better browser optimization.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx
Comment (miksansegundo) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx:178]: Only needs to run on mount

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx
Comment (miksansegundo) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-large-preview.tsx:198]: Moving `setDevice( device );` out of the timeout because the animation is triggered by it.

File: client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list-panel.tsx
Comment (miksansegundo) [inline client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list-panel.tsx:35]: No need for a `<p>` so better use `<div>`

**Review by arthur791004 — APPROVED** [review]
Tested and looks good to me 👍

[nit] Some changes might not be required as it just made the order different 😅

**Review by okmttdhr — APPROVED** [review]
LGTM✅

Worked well on Storybook also.

https://github.com/Automattic/wp-calypso/assets/5287479/bcbd6ca4-8928-4236-91f7-312cdf632951

**Comment by autumnfjeld** [comment]
@miksansegundo 
> We can address the scaling on another PR but I think they should be shipped together because 963 px or less is not enough for the desktop preview.

If the other PR is coming soon, go ahead and ship this one now. Because the improvements here are so good. Let's get these to our users with this PR. :)

**Comment by autumnfjeld** [comment]
Question: would it be worthwhile for us to add a track event somewhere in onboarding , maybe when the PA screen loads that tracks viewport widith?  We could something like 
`calypso_signup_viewportwidth = ` and use settings we use for responsive breakpoints. I'd need to find the code, is categorised as S M L ?

Or we could make our own designations:

`calypso_signup_assembler_viewportwidth = LT680 | 681to768| 769to1060 | 1061to1200 | 1201to1480 | GT1481`. I am just making up numbers. You get the idea :)

**Comment by arthur791004** [comment]
> Question: would it be worthwhile for us to add a track event somewhere in onboarding , maybe when the PA screen loads that tracks viewport widith?

I think it should be global property for all events as it's very useful 🙂

**Pull request merged**
